doc/scripts/gen_devicetree_rest.py

```diff
@@ -0,0 +1,654 @@
+    write_bindings_rst(vnd_lookup, out_dir)
+    write_per_compatible_orphans(bindings, base_binding, vnd_lookup, out_dir)
+
+def setup_bindings_dir(bindings, out_dir):
+    # Make a set of all the Path objects we will be creating for
+    # out_dir / bindings / {compatible}.rst. Delete all the ones that
+    # shouldn't be there. Make sure the bindings output directory
+    # exists.
+
+    paths = set()
+    bindings_dir = out_dir / 'bindings'
+    logger.info('making output subdirectory %s', bindings_dir)
+    bindings_dir.mkdir(parents=True, exist_ok=True)
+
+    for binding in bindings:
+        paths.add(bindings_dir / binding_filename(binding))
+
+    for path in bindings_dir.iterdir():
+        if path not in paths:
+            logger.info('removing unexpected file %s', path)
+            path.unlink()
+
+def write_bindings_rst(vnd_lookup, out_dir):
+    # Write out_dir / bindings.rst, the top level index of bindings.
+
+    string_io = io.StringIO()
+
+    print_block(f'''\
+    .. _devicetree_binding_index:
+
+    Bindings index
+    ##############
+
+    This page documents the available devicetree bindings.
+    See {zref('dt-bindings')} for an introduction to the Zephyr bindings
+    file format.
+
+    Vendor index
+    ************
+
+    This section contains an index of hardware vendors.
+    Click on a vendor's name to go to the list of bindings for
+    that vendor.
+
+    .. rst-class:: rst-columns
+    ''', string_io)
+
+    for vnd in vnd_lookup.vnd2bindings:
+        print(f'- :ref:`{vnd_lookup.target(vnd)}`', file=string_io)
+
+    print_block('''\
+
+    Bindings by vendor
+    ******************
+
+    This section contains available bindings, grouped by vendor.
+    Within each group, bindings are listed by the "compatible" property
+    they apply to, like this:
+
+    **Vendor name (vendor prefix)**
+
+    .. rst-class:: rst-columns
+
+    - <compatible-A>
+    - <compatible-B> (on <bus-name> bus)
+    - <compatible-C>
+    - ...
+
+    The text "(on <bus-name> bus)" appears when bindings may behave
+    differently depending on the bus the node appears on.
+    For example, this applies to some sensor device nodes, which may
+    appear as children of either I2C or SPI bus nodes.
+    ''', string_io)
+
+    for vnd, bindings in vnd_lookup.vnd2bindings.items():
+        if isinstance(vnd, tuple):
+            title = vnd[0]
+        else:
+            title = vnd_lookup.vendor(vnd).strip()
+            if isinstance(vnd, str):
+                title += f' ({vnd})'
+        underline = '=' * len(title)
+
+        print_block(f'''\
+        .. _{vnd_lookup.target(vnd)}:
+
+        {title}
+        {underline}
+
+        .. rst-class:: rst-columns
+        ''', string_io)
+        for binding in bindings:
+            print(f'- :ref:`{binding_ref_target(binding)}`', file=string_io)
+        print(file=string_io)
+
+    write_if_updated(out_dir / 'bindings.rst', string_io.getvalue())
```

Comment:
Shouldn't `os.path.join` be used here? I am sure eventually someone will want to build the documentation on Windows!